indispensable/src/commonMain/kotlin/at/asitplus/signum/indispensable/SymmetricEncryptionAlgorithm.kt

```diff
@@ -0,0 +1,128 @@
+package at.asitplus.signum.indispensable
+
+import at.asitplus.signum.indispensable.asn1.Identifiable
+import at.asitplus.signum.indispensable.asn1.KnownOIDs
+import at.asitplus.signum.indispensable.asn1.ObjectIdentifier
+import at.asitplus.signum.indispensable.mac.HMAC
+import at.asitplus.signum.indispensable.mac.MAC
+
+
+sealed interface SymmetricEncryptionAlgorithm<out A : AuthTrait> : Identifiable {
+    override fun toString(): String
+
+    companion object {
+
+        val AES_128 = AESDefinition(128u)
+        val AES_192 = AESDefinition(192u)
+        val AES_256 = AESDefinition(256u)
+
+        class AESDefinition(val keySize: UInt) {
+
+            val GCM = AES.GCM(keySize)
+            val CBC = CbcDefinition(keySize)
+
+            class CbcDefinition(keySize: UInt) {
+                val PLAIN = AES.CBC.Plain(keySize)
+                val HMAC = HmacDefinition(PLAIN)
+
+                class HmacDefinition(innerCipher: AES.CBC.Plain) {
+                    val SHA_256 = AES.CBC.HMAC(innerCipher, HMAC.SHA256)
+                    val SHA_384 = AES.CBC.HMAC(innerCipher, HMAC.SHA384)
+                    val SHA_512 = AES.CBC.HMAC(innerCipher, HMAC.SHA512)
+                    val SHA_1 = AES.CBC.HMAC(innerCipher, HMAC.SHA1)
+                }
+            }
+        }
+    }
+
+    val name: String
+
+
+    interface WithDedicatedMac : Authenticated, WithIV<AuthTrait.Authenticated> {
+        val mac: MAC
+        val innerCipher: SymmetricEncryptionAlgorithm.Unauthenticated
+    }
+
+    /**
+     * Indicates that a cipher requires an initialization vector
+     */
+    interface WithIV<A : AuthTrait> : SymmetricEncryptionAlgorithm<A> {
+        val ivNumBits: UInt
+    }
+
+    interface Authenticated : SymmetricEncryptionAlgorithm<AuthTrait.Authenticated>, AuthTrait.Authenticated
+    interface Unauthenticated : SymmetricEncryptionAlgorithm<AuthTrait.Unauthenticated>, AuthTrait.Unauthenticated
+
+    /**
+     * Key length in bits
+     */
+    val keyNumBits: UInt
+
+    sealed class AES<A : AuthTrait>(modeOfOps: ModeOfOperation, override val keyNumBits: UInt) :
+        BlockCipher<A>(modeOfOps, blockSizeBits = 128u) {
+        override val name: String = "AES-$keyNumBits ${modeOfOps.acronym}"
+
+        override fun toString(): String = name
+
+        class GCM internal constructor(keyNumBits: UInt) :
+            AES<AuthTrait.Authenticated>(ModeOfOperation.GCM, keyNumBits), WithIV<AuthTrait.Authenticated>,
+            Authenticated {
+            override val ivNumBits: UInt = 96u
+            override val tagNumBits: UInt = blockSizeBits
+            override val oid: ObjectIdentifier = when (keyNumBits) {
+                128u -> KnownOIDs.aes128_GCM
+                192u -> KnownOIDs.aes192_GCM
+                256u -> KnownOIDs.aes256_GCM
+                else -> throw IllegalStateException("$keyNumBits This is an implementation flaw. Report this bug!")
```

Comment:
don't we have `unreachable()` or similar for this

Reply:
not in this source tree.
Since we now have more cases where we run into illegal situations iff **we** messed up, we should probably introduce an exception specifically for that?!

Reply:
`IllegalStateException` (as thrown by `check`) is the correct exception

Reply:
(the stdlib should have `unreachable()` as an alias for `check(false)` but that's a different cup of tea)

Reply:
my bad.